Add VQE sample

guenp opened the pull request.

guenp commented:
We have this guidance for writing descriptions:

1. Length target of 85 characters (+/- 15 characters at the most)
2. Should not begin with “This sample” or other boilerplate wording
3. Should not mention the provider (as that is already noted on the tile)
4. Should speak in an implied 2nd-person view (“Run a job” preferred over “You can run a job”)
5. Verbiage should be specific to function and obvious if correlated to other samples in the gallery
6. Must be approved by a UX designer and PM

cgranade commented on samples/azure-quantum/variational-quantum-eigensolver/README.md:
```diff
@@ -0,0 +1,21 @@
+languages:
+- python
+products:
+- qiskit
```
I don't believe this is a valid product slug for the Microsoft-wide samples gallery. In particular, the Microsoft-wide docs team has only approved `azure-quantum` and `qdk` as quantum-related slugs at this time. The full list of valid product tags is listed at https://review.docs.microsoft.com/help/contribute/metadata-taxonomies?branch=main#product (internal-only).

guenp replied:
This is used in the getting-started sample, so I used it here for consistency.

guenp replied:
Ping @xinyi-joffre @achocron, what are your thoughts on this? Should qiskit be added as a valid product slug?

guenp replied:
We decided to remove the slug.

cgranade commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,520 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Remove idle wires\n",
```
It may help to explain why removing idle wires is useful here.

guenp replied:
Will do

cgranade commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,520 @@
+    "\n",
+    "circs = vqe._circuit_sampler._transpiled_circ_cache\n",
+    "circs = [remove_idle_qwires(circ) for circ in circs]\n",
+    "circ = circs[0]"
```
It feels a bit indirect to do a list comprehension over `circs`, but then to only keep one at the end?

guenp replied:
It was meant to offer the user a handle to both circuits if they wish to visualize them.

cgranade commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,520 @@
+   "id": "767fc7ea",
+   "metadata": {},
+   "source": [
+    "This shows the parametric trial state that is prepared and evaluated as part of VQE. The parameters, $\\theta[n]$, are assigned a value for each iteration."
```
```suggestion
    "This shows the parametric trial state that is prepared and evaluated as part of VQE. The parameters $\\theta[n]$ are assigned a value for each iteration."
```

cgranade commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,520 @@
+   "source": [
+    "## 2. Run on IonQ simulator via Azure Quantum\n",
+    "\n",
+    "Now, we can connect to Azure quantum and run VQE on the hardware backends."
```
```suggestion
    "Now, we can connect to the Azure Quantum service and run VQE on the hardware backends."
```

guenp replied:
I believe we call our product Azure Quantum.

cgranade replied:
Yes, the whole product, including the service, the Quantum Development Kit, and several other components.

guenp replied:
I just checked and according to PM:
> Our product is called "Azure Quantum"

guenp replied:
I suggest using "Azure Quantum Workspace" here to be syntactically correct.

cgranade replied:
I don't disagree that our product is called "Azure Quantum," but this sentence refers to a part of our product. For clarity and for consistency with our documentation, my comment remains as above.

cgranade reviewed (commented):
I've only reviewed the IonQ version so far, but in general I'd recommend a bit more explanatory text to help readers out along the way. Looking good, in any case!

guenp commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,520 @@
+   "source": [
+    "# Simulate the ground state of a Hydrogen molecule using Variational Quantum Eigensolver (VQE) on the IonQ simulator\n",
+    "\n",
+    "In this notebook, I'll show how to run VQE for a H2 molecule on the IonQ simulator via Azure Quantum using Qiskit.\n",
```
Zander recommends using the second person, so e.g. "In this notebook, you'll run ...".

guenp commented on samples/azure-quantum/variational-quantum-eigensolver/README.md:
```diff
@@ -0,0 +1 @@
+# Variational Quantum Eigensolver
```
Is the feature branch ahead of `main` or the other way around?

cgranade replied:
Neither is strictly ahead of the other at this point, which is not terribly optimal.

guenp commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,520 @@
+   "id": "47574490",
+   "metadata": {},
+   "source": [
+    "# Simulate the ground state of a Hydrogen molecule using Variational Quantum Eigensolver (VQE) on the IonQ simulator\n",
```
Because VQE can be quite expensive to run, this sample is meant for advanced users who already know what VQE is. It's not meant for absolute beginners - those folks should try out the getting started samples instead.

cgranade replied:
Agreed, that's my point; giving readers that we don't expect to get use out of the sample an easy exit to learn more is important when a sample is more advanced like that.

guenp commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,520 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qiskit.aqua.algorithms import VQE, NumPyEigensolver\n",
```
Sounds good.

guenp commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,520 @@
+   "id": "7dadd1bf",
+   "metadata": {},
+   "source": [
+    "First, prepare the qubit operators using PySCF."
```
I've added in a link to the package docs.

guenp commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,520 @@
+   "source": [
+    "driver = PySCFDriver(atom='H .0 .0 -0.3625; H .0 .0 0.3625', unit=UnitsType.ANGSTROM, charge=0, spin=0, basis='sto3g')\n",
+    "molecule = driver.run()\n",
+    "num_particles = molecule.num_alpha + molecule.num_beta\n",
```
I'll add an in-line comment as per Zander's suggestion.

guenp commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,520 @@
+   "id": "d3ceaa01",
+   "metadata": {},
+   "source": [
+    "### Simulate locally with noise and error mitigation"
```
I'll defer to the Qiskit documentation to explain that.

guenp commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,520 @@
+    "exact_solution = NumPyEigensolver(qubitOp).run()\n",
+    "print(\"Exact Result:\", np.real(exact_solution.eigenvalues) + molecule.nuclear_repulsion_energy)\n",
+    "optimizer = SPSA(maxiter=100)\n",
+    "var_form = EfficientSU2(qubitOp.num_qubits, entanglement=\"linear\")\n",
```
I'll add in a reference to the RyRz ansatz. As per the Qiskit documentation:
> As the effects of noise increase as the number of two qubit gates circuit depth increase, we use a heuristic variational form (RYRZ) rather than UCCSD as RYRZ has a much shallower circuit than UCCSD and uses substantially fewer two qubit gates.

guenp commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,520 @@
+    "driver = PySCFDriver(atom='H .0 .0 -0.3625; H .0 .0 0.3625', unit=UnitsType.ANGSTROM, charge=0, spin=0, basis='sto3g')\n",
+    "molecule = driver.run()\n",
+    "num_particles = molecule.num_alpha + molecule.num_beta\n",
+    "qubitOp = FermionicOperator(h1=molecule.one_body_integrals, h2=molecule.two_body_integrals).mapping(map_type='parity')\n",
```
Will do.

guenp commented:
@cgranade, I believe I have addressed all your comments, feel free to point it out if I forgot anything. Thanks!

lwoodyiii commented:
I could not get the first cell of the VQE-qiskit-hydrogen-ionq-sim to run. Got this error:

---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
/tmp/ipykernel_3897/4132631024.py in <module>
     12 from qiskit.aqua import QuantumInstance
     13 from qiskit.ignis.mitigation.measurement import CompleteMeasFitter
---> 14 from qiskit.providers.aer.noise import NoiseModel

ModuleNotFoundError: No module named 'qiskit.providers.aer'

<img width="784" alt="image" src="https://user-images.githubusercontent.com/1810351/166844926-2bfc58c6-4e76-4c69-bee3-ed0027bba6ff.png">

guenp commented:
Thanks, @lwoodyiii for pointing that out! It looks like a ton of my changes did not get picked up, perhaps something went wrong in one of the merges I did.
Could you take another look? Thanks again!

guenp commented:
![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/4041805/166979517-b4d579f2-7dba-435d-9d7f-e7d1c4754e51.png)

cgranade commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,564 @@
+   "id": "4d49d174",
+   "metadata": {},
+   "source": [
+    "## 2. Run on IonQ simulator via Azure Quantum Workspace\n",
```
```suggestion
    "## 2. Run on IonQ simulator via an Azure Quantum workspace\n",
```

cgranade commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,564 @@
+   "source": [
+    "## 2. Run on IonQ simulator via Azure Quantum Workspace\n",
+    "\n",
+    "Now, you can connect to the Azure Quantum Workspace and run VQE on the IonQ simulator backend."
```
```suggestion
    "Now, you can connect to the Azure Quantum workspace and run VQE on the IonQ simulator backend."
```

cgranade commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,564 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Connect to the Azure Quantum Workspace via a Qiskit provider\n",
```
```suggestion
    "# Connect to the Azure Quantum workspace via a Qiskit provider\n",
```

cgranade commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,564 @@
+   "id": "bc8a8b3e",
+   "metadata": {},
+   "source": [
+    "### Run one iteration on IonQ simulator via Azure Quantum\n",
```
```suggestion
    "### Run one iteration on IonQ simulator via the Azure Quantum service\n",
```

cgranade reviewed (commented):
I think this is getting a lot closer, thank you! I'd still suggest some minor edits for grammar, etc., but then it's just about ready to merge into both the portal gallery and main branches. Thanks!

guenp commented:
Thanks, @cgranade for the review. Please let me know in which branch/merge order you'd recommend to proceed.

achocron commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,564 @@
+   "source": [
+    "## 1. Simulate locally\n",
+    "\n",
+    "Here, you will simulate the program locally using the Aer simulator. You can create a `QuantumInstance` with a noise model, generated using a mock device `FakeVigo` with noise characteristics."
```
readability nit:
```suggestion
    "Here, you will simulate the program locally using the Aer simulator. You can create a `QuantumInstance` with a noise model using a mock device `FakeVigo` with noise characteristics."
```

achocron commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,564 @@
+   "source": [
+    "### Circuit visualization\n",
+    "\n",
+    "Each energy evaluation consists of two circuits that are run. You can visualize these circuits with Qiskit using the `vqe` instance."
```
nit
```suggestion
    "Each energy evaluation consists of two circuits. You can visualize these circuits with Qiskit using the `vqe` instance."
```

achocron commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,564 @@
+   "source": [
+    "# Connect to the Azure Quantum workspace via a Qiskit provider\n",
+    "from azure.quantum.qiskit import AzureQuantumProvider\n",
+    "provider = AzureQuantumProvider(\n",
```
Has this sample been tested in the hosted notebooks environment using the feature flag? I think that creating the IonQ backend must be in a different cell, as the login must be in its own cell. Can you please try to run it in the hosted environment and fix if necessary?

achocron commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,564 @@
+ },
+ "nbformat": 4,
+ "nbformat_minor": 5
+}
```
Can you please add a conclusion cell that wraps the notebook up nicely? It ends a little abruptly. What should I have learned from this notebook? What can I try next? Give the user a "high five" of sorts, too, because they got through this very cool example we should celebrate that!

achocron commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-quantinuum-emulator.ipynb:
```diff
@@ -0,0 +1,717 @@
+   "source": [
+    "## 1. Simulate locally\n",
+    "\n",
+    "Here, you will simulate the program locally using the Aer simulator. You can create a `QuantumInstance` with a noise model, generated using a mock device `FakeVigo` with noise characteristics."
```
nit
```suggestion
    "Here, you will simulate the program locally using the Aer simulator. You can create a `QuantumInstance` with a noise model using a mock device `FakeVigo` with noise characteristics."
```

achocron commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-quantinuum-emulator.ipynb:
```diff
@@ -0,0 +1,717 @@
+   "outputs": [],
+   "source": [
+    "# Connect to the Azure Quantum Workspace via a Qiskit provider\n",
+    "from azure.quantum.qiskit import AzureQuantumProvider\n",
```
same comment as in other notebook

xinyi-joffre commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-ionq-sim.ipynb:
```diff
@@ -0,0 +1,564 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install qiskit-aer"
```
Thanks, I'll see if I can add this to the base image!

xinyi-joffre commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-quantinuum-emulator.ipynb:
```diff
@@ -0,0 +1,717 @@
+    "\n",
+    "![Hydrogen molecule](https://user-images.githubusercontent.com/4041805/166981145-c33b8d1a-24d1-4776-91ee-f514b0a5ab04.jpg)\n",
+    "\n",
+    "In this notebook, you'll learn how to run VQE for a $H_{2}$ molecule on the IonQ simulator using Qiskit on an Azure Quantum backend.\n",
```
```suggestion
    "In this notebook, you'll learn how to run VQE for a $H_{2}$ molecule on the Quantinuum emulator using Qiskit on an Azure Quantum backend.\n",
```

xinyi-joffre commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-quantinuum-emulator.ipynb:
```diff
@@ -0,0 +1,717 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "cost = [ionq_qpu_backend.estimate_cost(circ, shots=1000) for circ in ready_circs]"
```
It looks like this cell is for ionq. Should we remove from this sample?

xinyi-joffre commented on samples/azure-quantum/variational-quantum-eigensolver/VQE-qiskit-hydrogen-quantinuum-emulator.ipynb:
```diff
@@ -0,0 +1,717 @@
+   ],
+   "source": [
+    "for _cost in cost:\n",
+    "    print(_cost.estimated_total, _cost.currency_code)"
```
Same comment as above, I think this cell is currently printing IonQ costs, but we can probably remove these two cells from this sample!

guenp commented:
Thanks @xinyi-joffre and @achocron, I believe I have addressed all your comments! Do you want to do a final check to make sure it's ready to be merged?
